crates/btc-relay/src/lib.rs

```diff
@@ -663,26 +627,82 @@ impl<T: Config> Pallet<T> {
         Ok(())
     }
 
-    pub fn _verify_and_validate_transaction(
-        raw_merkle_proof: Vec<u8>,
+    /// helper for the dispatchable
+    fn _validate_transaction(
         raw_tx: Vec<u8>,
+        expected_btc: i64,
         recipient_btc_address: BtcAddress,
-        minimum_btc: Option<i64>,
         op_return_id: Option<H256>,
-        confirmations: Option<u32>,
+    ) -> Result<(), DispatchError> {
+        let transaction = Self::parse_transaction(&raw_tx)?;
+        match op_return_id {
+            Some(op_return) => {
+                Self::validate_op_return_transaction(transaction, recipient_btc_address, expected_btc, op_return)?;
+            }
+            None => {
+                let payment = Self::get_issue_payment(transaction, recipient_btc_address)?;
+                ensure!(payment.1 == expected_btc, Error::<T>::InvalidPaymentAmount);
+            }
+        };
+        Ok(())
+    }
+
+    /// interface to the issue pallet; verifies inclusion, and returns the first input
+    /// address (for refunds) and the payment amount
+    pub fn get_and_verify_issue_payment(
+        raw_merkle_proof: Vec<u8>,
+        raw_tx: Vec<u8>,
+        recipient_btc_address: BtcAddress,
+    ) -> Result<(BtcAddress, i64), DispatchError> {
+        let transaction = Self::parse_transaction(&raw_tx)?;
+
+        // Verify that the transaction is indeed included in the main chain
+        Self::_verify_transaction_inclusion(transaction.tx_id(), raw_merkle_proof, None)?;
+
+        Self::get_issue_payment(transaction, recipient_btc_address)
+    }
+
+    fn get_issue_payment(
```

Comment:
Would prefer `Option<H256>` for generality over specific issue / redeem / replace functions

Reply:
I specifically got away from the generality - imo it didn't make sense to generalize, when the input, the output, and the actual checks are all different between issue and [redeem,replace,refund]:



| 			| issue							| redeem/replace/refund                                                              |
| ------------- |------------| -----|
| input:		| address						| 	address, amount, op_return                                                   |
| returns:	| origin_address, paid amount 	| return-to-self address (theft check)                                                             |
| checks:		| any payment to _address_		| payment of correct amount to _address_, exactly 1 correct op_return, at most 1 return-to-self, no unparsable outputs

Reply:
Okay that makes sense